backend/junior-3/haskell.md

```diff
@@ -197,17 +197,18 @@
   * `1:(thunk)`
   * `1:2:(thunk)`
   * `1:2:3:[]`
+  * `thunk:thunk`
   * `15`
   * `\x -> x * 2`
   * `(\x -> x + 1) 3`
 * Can haskell evaluate in strict mode?
-* Why strict functions in Haskell evaluate values to WHNF and not NF?
 * What is the function `seq` (and operator `$!`)?
 * What is the function `deepseq` (and operator `$!!`)?
 * Could using `seq` change the returned value of the function?
+* What is the expression result: ``(a + b) `seq` ((a + b) : list)``? Why?
 * What is the GHC extension `BangPatterns`?
   * Make examples when bang pattern is useless.
-  * Make examples when bang pattern has less power, than it could be supposed.
```

Comment:
Где-то здесь нужно рассказать про такие факты:
1. Банг-паттерн внутри лет-биндинга становится таким же ленивым, как и обычный лет-биндинг:
```hs
try @SomeException (evaluate (let (!x, y) = error "urk" in 10)) >>= print -- Right 10
```
2. Вообще, банг-паттерн задаёт не время вычисления, а зависимость одного выражения от другого. Если результат не форсируется - то и бэнг ничего не сделает. Например:
```hs
try @SomeException (evaluate (let !x = error "urk" in 10, 20)) >>= print . fmap snd -- Right 20
```
Мне кажется, оба этих примера подходят под этот вопрос, и оба случая следует иметь в виду при использовании расширения - так что, если этот вопрос убирается, то вместо него следует написать что-то другое, которое спросит именно про эти нюансы.

Reply:
На сдаче мне сказали, что на этот вопрос подразумевался ответ "бэнги вычисляют до WHNF, не до NF". Имхо, вопрос про `less power than it could be supposed` не очень правильный, потому что если человек читал доки внимательно, то ему не придет в голову, что bang pattern форсирует до NF :). А вопрос про то, почему хаскель любит WHNF, а не NF, у нас уже есть.

Про то, что ты предлагаешь, ниже есть вопрос. Кажется, тут спрошено все что нужно:

> Do bang patterns force execution when they are nested in constructors inside
    `let` or `where` expressions? Does this force `a` to be evaluated and when
    it does?
>
>    let (Just !a) = x in 1 + 1

Reply:
А нет, здесь не все. Я еще добавлю пример про что-нибудь вроде `let (Just !a) = [a]`, надо будет ответить, когда вычислится `a`.

Reply:
Готово